benchexec/resources.py

```diff
@@ -15,313 +15,916 @@
+    such that the assigned cores are a subset of the cores that the current process is allowed to use.
+    Furthermore all currently supported topology data is read for each core and
+    the cores are then organised accordingly into hierarchy_levels.
+    hierarchy_levels is sorted so that the first dict maps hyper-threading siblings
+    while the next dict in the list subsumes same or more cores per key (topology identifier)
+    as the siblings dict but less than the following dict.
+    Therefore when iterating through the list of dicts, each dict has less keys
+    but the corresponding value is a list of greater length than the previous dict had.
+    Thus hierarchy_levels reflects a hierarchy of the available topology layers from smallest to largest.
+    Additionally, the list of available cores is converted into a list of VirtualCore objects
+    that provide its ID and a list of the memory regions it belongs to.
 
-    @param coreLimit: the number of cores for each run
-    @param num_of_threads: the number of parallel benchmark executions
-    @param coreSet: the list of CPU cores identifiers provided by a user, None makes benchexec using all cores
-    @return a list of lists, where each inner list contains the cores for one run
+    This script does currently not support situations where the available cores are
+    asymmetrically split over CPUs, e.g. 3 cores on one CPU and 5 on another.
+
+    @param coreLimit:           the number of cores for each thread
+    @param num_of_threads:      the number of parallel benchmark executions
+    @param use_hyperthreading:  boolean to check if no-hyperthreading method is being used
+    @param coreSet:             the list of CPU core identifiers provided by a user,None makes benchexec using all cores
+    @return list of lists, where each inner list contains the cores for one run
     """
+
+    if (
+        type(coreLimit) != int
+        or type(num_of_threads) != int
+        or type(use_hyperthreading) != bool
+    ):
+        sys.exit("Incorrect data type entered")
+
+    if coreLimit < 1 or num_of_threads < 1:
+        sys.exit("Only integers > 0 accepted for coreLimit & num_of_threads")
+
+    hierarchy_levels = []
     try:
-        # read list of available CPU cores
-        allCpus = my_cgroups.read_allowed_cpus()
-
-        # Filter CPU cores according to the list of identifiers provided by a user
-        if coreSet:
-            invalid_cores = sorted(set(coreSet).difference(set(allCpus)))
-            if len(invalid_cores) > 0:
-                raise ValueError(
-                    "The following provided CPU cores are not available: "
-                    + ", ".join(map(str, invalid_cores))
-                )
-            allCpus = [core for core in allCpus if core in coreSet]
+        # read list of available CPU cores (int)
+        allowedCpus = get_cpu_list(my_cgroups)
+        allCpus_list = get_cpu_list(my_cgroups, coreSet)
```

Comment:
I notice that this function is called twice, and think that this will lead to highly confusing log output, because the methods logs all cores, and so there will be two log messages with potentially different sets of cores, both claiming to be the available cores. Additionally, there are the log messages by the frequency filter.

I think this should be refactored such that the cores are read just once, logged just once, and the frequency filter is applied and logged just once.